File: HW2/01.py
cntin = 0
cntpre = 0
cntpost = 0

class Node:
    def __init__(self, k):
        self.left = None
        self.right = None
        self.val = int(k)
 
def insert(root, k):
    if root is None:
        return Node(int(k))
    else:
        if root.val == k:
            return root
        elif root.val < k:
            root.right = insert(root.right, int(k))
        else:
            root.left = insert(root.left, int(k))
    return root 
 
def inorder(root, n):
    global cntin
    if root:
        inorder(root.left, n)
        if(cntin < n - 1):
            print(root.val, end=' ')
            cntin += 1
        else:
            print(root.val)
        inorder(root.right, n)

def preorder(root, n):
    global cntpre
    if root:
        if(cntpre < n - 1):
            print(root.val, end = ' ')
            cntpre += 1
        else:
           print(root.val)
        preorder(root.left, n)
        preorder(root.right, n)

def postorder(root, n):
    global cntpost
    if root:
        postorder(root.left, n)
        postorder(root.right, n)
        if(cntpost < n - 1):
            print(root.val, end = ' ')
            cntpost += 1
        else:
            print(root.val)

# main
a = list(map(int, input().split()))

r = Node(a[0])
for i in (a):
    if i != 0:
        r = insert(r, int(i))

inorder(r, len(a))
preorder(r, len(a))
postorder(r, len(a))

File: HW2/01.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>

using namespace std;

class Tree
{
public:
    int val;
    Tree *left;
    Tree *right;
    Tree()
    {
        val = 0;
        left = NULL;
        right = NULL;
    }
    Tree(int n)
    {
        val = n;
        left = right = NULL;
    }

    Tree* insert(Tree* rt, int n)
    {
        if(!rt)
            return new Tree(n);
        
        if(n < rt->val)
            rt->left = insert(rt->left, n);

        else if(n > rt->val)
            rt->right = insert(rt->right, n);
        
        return rt;
    }

    void inorder(Tree* rt)
    {
        if(!rt)
            return;
        inorder(rt->left);
        cout << rt->val << " ";
        inorder(rt->right);
    }

    void preorder(Tree* rt)
    {
        if(!rt)
            return;
        cout << rt->val << " ";
        preorder(rt->left);
        preorder(rt->right);
    }

    void postorder(Tree* rt)
    {
        if(!rt)
            return;
        postorder(rt->left);
        postorder(rt->right);
        cout << rt->val << " ";
    }
};

int main()
{
    Tree t, *r = NULL;
    string s;
    stringstream ss;
    vector<int> num;

    getline(cin, s);
    ss << s;

    string tmp;
    int found;

    while (!ss.eof())
    {
        ss >> tmp;
        if(stringstream(tmp) >> found)
            num.push_back(found);
        tmp = "";
    }

    r = t.insert(r, num[0]);
    for (int i = 1; i < num.size(); i++)
        t.insert(r, num[i]);
    

    t.inorder(r);
    cout << endl;
    t.preorder(r);
    cout << endl;
    t.postorder(r);
}